src/compiler/checker.ts

```diff
@@ -8987,6 +8978,45 @@ module ts {
             }
         }
 
+        function checkElementTypeOfArrayOrString(arrayOrStringType: Type, expressionForError: Expression): Type {
+            Debug.assert(languageVersion < ScriptTarget.ES6);
+            var isJustString = allConstituentTypesHaveKind(arrayOrStringType, TypeFlags.StringLike);
+            
+            // Check isJustString because removeTypesFromUnionType will only remove types if it doesn't result

```

Comment:
Yeah I think you are right. I was actually thinking about that when I did this. I can add a parameter called allowEmptyUnionResult.